File: core/src/main/java/cloudspec/annotation/ResourceReflectionUtil.java
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.annotation;

import cloudspec.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Util to extract resource information out of annotated classes.
 */
public class ResourceReflectionUtil {
    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceReflectionUtil.class);

    /**
     * Extract resource out of an object of a class annotated with @ResourceDefinition.
     *
     * @param obj Object.
     * @return Optional resource.
     */
    public static Optional<Resource> toResource(Object obj) {
        return ResourceDefReflectionUtil.toResourceDefRef(obj.getClass())
                .flatMap(resourceDefRef -> toResourceId(obj)
                        .map(resourceId -> new Resource(
                                resourceDefRef,
                                resourceId,
                                ResourceReflectionUtil.toProperties(obj),
                                ResourceReflectionUtil.toAssociations(obj)
                        )));
    }

    /**
     * Extract resource id out of an object with a class annotated with @ResourceDefinition.
     *
     * @param obj Object.
     * @return Optional resource id.
     */
    public static Optional<String> toResourceId(Object obj) {
        return Stream.of(obj.getClass().getDeclaredFields())
                .peek(field -> {
                    if (!field.isAnnotationPresent(IdDefinition.class)) {
                        LOGGER.debug(
                                "Cannot produce a resource id from field '{}' of object of class '{}' " +
                                        "because it does not have a field annotated with @IdDefinition",
                                field.getName(),
                                obj.getClass().getCanonicalName()
                        );
                    }
                })
                .filter(field -> field.isAnnotationPresent(IdDefinition.class))
                .filter(field -> field.getType().isAssignableFrom(String.class))
                .map(field -> {
                    if (LOGGER.isDebugEnabled()) {
                        LOGGER.debug("Found @IdDefinition annotation in field '{}' of object of class '{}'",
                                field.getName(),
                                obj.getClass().getCanonicalName()
                        );
                    }

                    try {
                        field.setAccessible(true);
                        String id = (String) field.get(obj);
                        field.setAccessible(true);

                        return id;
                    } catch (IllegalAccessException exception) {
                        throw new RuntimeException(
                                String.format(
                                        "Could not obtain ID from field '%s' in class '%s'",
                                        field.getName(),
                                        obj.getClass().getCanonicalName()
                                )
                        );
                    }
                }).findFirst();

    }

    /**
     * Extract all properties out of fields annotated with @PropertyDefinition.
     *
     * @param obj Object.
     * @return List of properties.
     */
    public static Properties toProperties(Object obj) {
        return new Properties(
                Stream.of(obj.getClass().getDeclaredFields())
                        .filter(field -> field.isAnnotationPresent(PropertyDefinition.class))
                        .flatMap(field -> {
                            if (LOGGER.isDebugEnabled()) {
                                LOGGER.debug("Found @PropertyDefinition annotation in field '{}' of class '{}'",
                                        field.getName(),
                                        field.getType().getCanonicalName()
                                );
                            }

                            PropertyDefinition propertyDefAnnotation = field.getAnnotation(PropertyDefinition.class);

                            try {
                                field.setAccessible(true);
                                Object value = field.get(obj);
                                Boolean multiValued = ResourceDefReflectionUtil.guessMultiValued(field);
                                field.setAccessible(true);

                                switch (ResourceDefReflectionUtil.guessPropertyType(field)) {
                                    case INTEGER:
                                    case DOUBLE:
                                    case STRING:
                                    case BOOLEAN:
                                    case KEY_VALUE:
                                        return Stream.of(new Property(propertyDefAnnotation.name(), value));
                                    case NESTED:
                                    default:
                                        if (multiValued) {
                                            return Stream.of(
                                                    new Property(
                                                            propertyDefAnnotation.name(),
                                                            ((List<?>) value)
                                                                    .stream()
                                                                    .map(ResourceReflectionUtil::toProperties)
                                                                    .collect(Collectors.toList())
                                                    )
                                            );
                                        }

                                        Properties properties = toProperties(value);
                                        Associations associations = toAssociations(value);
                                        if (properties.size() > 0 || associations.size() > 0) {
                                            return Stream.of(
                                                    new Property(
                                                            propertyDefAnnotation.name(),
                                                            new Properties(properties, associations)
                                                    )
                                            );
                                        }
                                }

                                LOGGER.warn("Type {} of property '{}' in class {} is not supported",
                                        field.getType().getCanonicalName(),
                                        field.getName(),
                                        obj.getClass().getCanonicalName()
                                );
                                return Stream.empty();
                            } catch (IllegalAccessException exception) {
                                throw new RuntimeException(
                                        String.format(
                                                "Could not obtain value from property %s in class %s",
                                                field.getName(),
                                                obj.getClass().getCanonicalName()
                                        )
                                );
                            }
                        })
        );
    }

    /**
     * Extract all associations out of fields annotated with @AssociationDefinition.
     *
     * @param obj Object.
     * @return List of associations.
     */
    public static Associations toAssociations(Object obj) {
        return new Associations(
                Stream.of(obj.getClass().getDeclaredFields())
                        .filter(field -> field.isAnnotationPresent(AssociationDefinition.class))
                        .flatMap(field -> {
                            // validate field is a string
                            if (!field.getType().isAssignableFrom(String.class)) {
                                LOGGER.warn("Type {} of association '{}' in class {} is not supported",
                                        field.getType().getCanonicalName(),
                                        field.getName(),
                                        obj.getClass().getCanonicalName()
                                );
                                return Stream.empty();
                            }

                            AssociationDefinition associationDefinitionAnnotation = field.getAnnotation(AssociationDefinition.class);

                            String associationName = associationDefinitionAnnotation.name();
                            Optional<ResourceDefRef> resourceDefRefOptional = ResourceDefReflectionUtil.toResourceDefRef(associationDefinitionAnnotation.targetClass());

                            // TODO add validation for association name and resourceDefRef
                            if (!resourceDefRefOptional.isPresent()) {
                                return Stream.empty();
                            }

                            try {
                                field.setAccessible(true);
                                String id = (String) field.get(obj);
                                field.setAccessible(true);

                                return Stream.of(new Association(
                                        associationName,
                                        resourceDefRefOptional.get(),
                                        id)
                                );
                            } catch (IllegalAccessException exception) {
                                throw new RuntimeException(
                                        String.format(
                                                "Could not obtain value from property %s in class %s",
                                                field.getName(),
                                                obj.getClass().getCanonicalName()
                                        )
                                );
                            }
                        })
        );
    }
}

File: core/src/main/kotlin/cloudspec/annotation/ResourceReflectionUtil.kt
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.annotation

import arrow.core.Option
import arrow.core.Some
import arrow.core.firstOrNone
import cloudspec.annotation.ResourceDefReflectionUtil.guessContainedType
import cloudspec.annotation.ResourceDefReflectionUtil.guessPropertyType
import cloudspec.annotation.ResourceDefReflectionUtil.isNullableString
import cloudspec.annotation.ResourceDefReflectionUtil.isString
import cloudspec.annotation.ResourceDefReflectionUtil.toResourceDefRef
import cloudspec.model.*
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.*
import kotlin.reflect.full.findAnnotation
import kotlin.reflect.full.memberProperties

/**
 * Util to extract resource information out of annotated classes.
 */
object ResourceReflectionUtil {
    private val logger = LoggerFactory.getLogger(ResourceReflectionUtil::class.java)

    /**
     * Extract resource out of an object of a class annotated with @ResourceDefinition.
     *
     * @param obj Object.
     * @return Optional resource.
     */
    fun toResource(obj: Any): Option<Resource> {
        return toResourceDefRef(obj::class).flatMap { ref ->
            toResourceId(obj).map { id ->
                Resource(ref, id, toProperties(obj), toAssociations(obj))
            }
        }
    }

    /**
     * Extract resource id out of an object with a class annotated with @ResourceDefinition.
     *
     * @param obj Object.
     * @return Resource id or null.
     */
    fun toResourceId(obj: Any): Option<String> {
        return obj::class.memberProperties
                .onEach {
                    if (it.findAnnotation<IdDefinition>() == null || !isString(it.returnType)) {
                        logger.debug("Cannot produce a resource id from property '${it.name}' " +
                                             "of object of class '${obj::class.qualifiedName}' " +
                                             "because it is not annotated with @IdDefinition " +
                                             "or is not a String")
                    }
                }
                .filter { it.findAnnotation<IdDefinition>() != null && isString(it.returnType) }
                .map {
                    if (logger.isDebugEnabled) {
                        logger.debug("Found @IdDefinition annotation in property '${it.name}' " +
                                             "of object of class '${obj::class.qualifiedName}'")
                    }

                    try {
                        return@map it.call(obj) as String
                    } catch (exception: IllegalAccessException) {
                        throw RuntimeException("Could not obtain ID from property '${it.name}' " +
                                                       "of object of class ${obj::class.qualifiedName}'")
                    }
                }
                .firstOrNone()
    }

    /**
     * Extract all properties out of object properties annotated with @PropertyDefinition.
     *
     * @param obj Object.
     * @return Set of properties.
     */
    fun toProperties(obj: Any): Set<Property<*>> {
        return obj::class.memberProperties
                .filter { it.findAnnotation<PropertyDefinition>() != null }
                .flatMap { kprop ->
                    val propertyDefAnnotation = kprop.findAnnotation<PropertyDefinition>()!!

                    if (logger.isDebugEnabled) {
                        logger.debug("Found @PropertyDefinition annotation in property '${kprop.name}' " +
                                             "of object of class '${kprop.returnType}'")
                    }

                    try {
                        // get property value in set form
                        val values = when (val value = kprop.call(obj)) {
                            is List<*> -> value.toSet()
                            is Set<*> -> value
                            else -> setOf(value)
                        }

                        values.mapNotNull { value ->
                            when (val propertyType = guessPropertyType(kprop)) {
                                is Some -> when (propertyType.t) {
                                    PropertyType.NUMBER ->
                                        NumberProperty(propertyDefAnnotation.name, (value as Number?))
                                    PropertyType.STRING ->
                                        StringProperty(propertyDefAnnotation.name, (value as String?))
                                    PropertyType.BOOLEAN ->
                                        BooleanProperty(propertyDefAnnotation.name, (value as Boolean?))
                                    PropertyType.DATE ->
                                        InstantProperty(
                                                propertyDefAnnotation.name,
                                                when (value) {
                                                    is Date -> value.toInstant()
                                                    is Instant -> value
                                                    else -> null
                                                }
                                        )
                                    PropertyType.KEY_VALUE ->
                                        KeyValueProperty(propertyDefAnnotation.name, (value as KeyValue?))
                                    PropertyType.NESTED ->
                                        NestedProperty(
                                                propertyDefAnnotation.name,
                                                value?.let {
                                                    NestedPropertyValue(toProperties(value), toAssociations(value))
                                                }
                                        )
                                }
                                else -> {
                                    val realType = guessContainedType(kprop.returnType)
                                    logger.warn("Type $realType of property '${kprop.name}' " +
                                                        "in class ${obj::class.qualifiedName} is not supported")
                                    null
                                }
                            }
                        }
                    } catch (exception: IllegalAccessException) {
                        throw RuntimeException("Could not obtain value from property ${kprop.name} " +
                                                       "in class ${obj::class.qualifiedName}")
                    }
                }
                .toSet()
    }

    /**
     * Extract all associations out of object properties annotated with @AssociationDefinition.
     *
     * @param obj Object.
     * @return List of associations.
     */
    fun toAssociations(obj: Any): Set<Association> {
        return obj::class.memberProperties
                .filter { it.findAnnotation<AssociationDefinition>() != null }
                .flatMap { kprop ->
                    val associationDefinitionAnnotation = kprop.findAnnotation<AssociationDefinition>()!!

                    // get type
                    val containedType = guessContainedType(kprop.returnType)

                    // only strings are supported
                    if (!isNullableString(containedType)) {
                        logger.warn("Type $containedType of association '${kprop.name}' " +
                                            "in class ${obj::class.qualifiedName} is not supported")
                        return@flatMap emptyList<Association>()
                    }

                    when (val resourceDefRefOpt= toResourceDefRef(associationDefinitionAnnotation.targetClass)) {
                        is Some -> {
                            // TODO add validation for association name and resourceDefRef
                            try {
                                kprop.call(obj)?.let { id ->
                                    when (id) {
                                        is List<*> -> id.toSet()
                                        is Set<*> -> id
                                        else -> setOf(id)
                                    }.map { resourceId ->
                                        Association(associationDefinitionAnnotation.name,
                                                    resourceDefRefOpt.t,
                                                    resourceId as String)
                                    }
                                } ?: emptyList()
                            } catch (exception: IllegalAccessException) {
                                throw RuntimeException("Could not obtain value from property ${kprop.name} " +
                                                               "in class ${obj::class.qualifiedName}")
                            }
                        }
                        else -> emptyList()
                    }
                }
                .toSet()
    }
}
